packages/di/src/utils.ts

```diff
@@ -1,44 +1,49 @@
-import {
-  ServiceIdentifier,
-  Provider,
-  Type,
-  ValueProvider,
-  ClassProvider,
-  FactoryProvider,
-  TypeProvider,
-} from './types';
-
-export const DESIGN_PARAMTYPES = 'design:paramtypes';
-export const DESIGN_TYPE = 'design:type';
-export const PROVIDER_OPTIONS = 'provider-options';
-export const PROPERTY_KEYS = 'property-keys';
-
-export function getServiceIdentifierName<T>(serviceIdentifier: ServiceIdentifier<T>) {
-  if (typeof serviceIdentifier === 'function' && isType<T>(serviceIdentifier)) {
-    return serviceIdentifier.name;
-  } else if (typeof serviceIdentifier !== 'undefined') {
-    return serviceIdentifier.toString();
-  } else {
-    return 'undefined';
-  }
+export const ERROR_TYPE = 'diType';
+export const ERROR_ORIGINAL_ERROR = 'diOriginalError';
+export const ERROR_LOGGER = 'diErrorLogger';
+
+export function getType(error: Error): Function {
+  return (error as any)[ERROR_TYPE];
 }
 
-export function isType<T>(v: any): v is Type<T> {
-  return typeof v === 'function' && 'prototype' in v;
+export function getOriginalError(error: Error): Error {
+  return (error as any)[ERROR_ORIGINAL_ERROR];
 }
 
-export function isTypeProvider<T>(v: Provider<T>): v is TypeProvider<T> {
-  return isType<T>(v);
+function defaultErrorLogger(console: Console, ...values: any[]) {
+  (<any>console.error)(...values);
 }
 
-export function isValueProvider<T>(v: Provider<T>): v is ValueProvider<T> {
-  return 'useValue' in v;
+export function getErrorLogger(
+  error: Error
+): (console: Console, ...values: any[]) => void {
+  return (error as any)[ERROR_LOGGER] || defaultErrorLogger;
 }
 
-export function isClassProvider<T>(v: Provider<T>): v is ClassProvider<T> {
-  return 'useClass' in v && isType(v.useClass);
+export function wrappedError(message: string, originalError: any): Error {
+  const msg = `${message} caused by: ${
+    originalError instanceof Error ? originalError.message : originalError
+  }`;
+  const error = Error(msg);
+  (error as any)[ERROR_ORIGINAL_ERROR] = originalError;
+  return error;
 }
 
-export function isFactoryProvider<T>(v: Provider<T>): v is FactoryProvider<T> {
-  return 'useFactory' in v && typeof v.useFactory === 'function';
+export function stringify(token: any): string {
+  if (typeof token === 'string') {
+    return token;
+  }
+
+  if (token == null) {
+    return '' + token;
```

Comment:
This is done in order to return `'null'`? maybe just returns it? :D

Reply:
Also, what about `undefined`? this will eventually get to `token.name` and fail, no